samples/msal-node-samples/client-credentials-distributed-cache/src/AuthProvider.ts

```diff
@@ -0,0 +1,148 @@
+
+import CustomCachePlugin from "./CustomCachePlugin";
+import RedisClientWrapper from "./RedisClientWrapper";
+import AxiosHelper from './AxiosHelper';
+
+export type AppConfig = {
+    instance: string;
+    tenantId: string;
+    clientId: string;
+    clientSecret: string;
+};
+
+export class AuthProvider {
+    private msalConfig: Configuration;
+    private cacheClientWrapper: RedisClientWrapper;
+
+    private constructor(msalConfig: Configuration, cacheClient: RedisClientType) {
+        this.msalConfig = msalConfig;
+        this.cacheClientWrapper = new RedisClientWrapper(cacheClient);
+    }
+
+    static async initialize(appConfig: AppConfig, cacheClient: RedisClientType): Promise<AuthProvider> {
+        const msalConfig = {
+            auth: {
+                clientId: appConfig.clientId,
+                authority: `${appConfig.instance}/${appConfig.tenantId}`,
+                clientSecret: appConfig.clientSecret,
+            },
+            system: {
+                loggerOptions: {
+                    loggerCallback(loglevel, message, containsPii) {
+                        console.log(message);
+                    },
+                    logLevel: LogLevel.Trace,
+                    piiLoggingEnabled: false
+                },
+                // proxyUrl: "http://localhost:8888" // uncomment to capture traffic with Fiddler
+            }
+        } as Configuration;
+
+        const msalConfigWithMetadata = await AuthProvider.getMetadata(msalConfig, cacheClient);
+
+        return new AuthProvider(msalConfigWithMetadata, cacheClient);
+    }
+
+    async getToken(tokenRequest: ClientCredentialRequest): Promise<AuthenticationResult | null> {
+        const cca = new ConfidentialClientApplication({
+            ...this.msalConfig,
+            cache: {
+                cachePlugin: new CustomCachePlugin(
+                    this.cacheClientWrapper,
+                    `${this.msalConfig.auth.clientId}.${this.msalConfig.auth.authority!.split("/").pop()!}` // partitionKey <clientId>.<tenantId>
+                )
+            }
+        });
+
+        let tokenResponse = null;
+
+        try {
+            performance.mark("acquireTokenByClientCredential-start");
+            await cca.getTokenCache().getAllAccounts(); // required for triggering cache read
+            tokenResponse = await cca.acquireTokenByClientCredential(tokenRequest);
+            performance.mark("acquireTokenByClientCredential-end");
+            performance.measure(
+                tokenResponse?.fromCache ? "acquireTokenByClientCredential-fromCache" : "acquireTokenByClientCredential-fromNetwork",
+                "acquireTokenByClientCredential-start",
+                "acquireTokenByClientCredential-end"
+            );
+        } catch (error) {
+            throw error;
+        }
+
+        return tokenResponse;
+    }
+
+    private static async getMetadata(msalConfig: Configuration, cacheClient: RedisClientType): Promise<Configuration> {
```

Comment:
Add a comment explaining this "optimization" and please make sure you capture this in a work item. Ideally we'd have a work item like "Make MSAL multi-tenant S2S scenario scalable out of the box" with all the things that we need to fix in MSAL so that this sample can be reduced from 100s of lines to <50.

Reply:
comment added with a link to the perf doc https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/performance.md